libp2p/src/main/kotlin/io/libp2p/security/tls/TLSSecureChannel.kt

```diff
@@ -0,0 +1,315 @@
+import java.security.interfaces.ECPublicKey
+import java.security.spec.*
+import java.time.Instant
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.logging.Level
+import java.util.logging.Logger
+import javax.net.ssl.X509TrustManager
+
+private val log = Logger.getLogger(TlsSecureChannel::class.java.name)
+private val SetupHandlerName = "TlsSetup"
+val certificatePrefix = "libp2p-tls-handshake:".encodeToByteArray()
+
+class TlsSecureChannel(private val localKey: PrivKey, private val muxerIds: List<String>, private val certAlgorithm: String) :
+    SecureChannel {
+
+    constructor(localKey: PrivKey, muxerIds: List<String>) : this(localKey, muxerIds, "Ed25519") {}
+
+    companion object {
+        const val announce = "/tls/1.0.0"
+        init {
+            Security.insertProviderAt(Libp2pCrypto.provider, 1)
+            Security.insertProviderAt(BouncyCastleJsseProvider(), 2)
+        }
+    }
+
+    override val protocolDescriptor = ProtocolDescriptor(announce)
+
+    fun initChannel(ch: P2PChannel): CompletableFuture<SecureChannel.Session> {
+        return initChannel(ch, "")
+    }
+
+    override fun initChannel(
+        ch: P2PChannel,
+        selectedProtocol: String
+    ): CompletableFuture<SecureChannel.Session> {
+        val handshakeComplete = CompletableFuture<SecureChannel.Session>()
+        ch.pushHandler(SetupHandlerName, ChannelSetup(localKey, muxerIds, certAlgorithm, ch, handshakeComplete))
+        return handshakeComplete
+    }
+}
+
+fun buildTlsHandler(
+    localKey: PrivKey,
+    expectedRemotePeer: Optional<PeerId>,
+    muxerIds: List<String>,
+    certAlgorithm: String,
+    ch: P2PChannel,
+    handshakeComplete: CompletableFuture<SecureChannel.Session>,
+    ctx: ChannelHandlerContext
+): SslHandler {
+    val connectionKeys = if (certAlgorithm.equals("ECDSA")) generateEcdsaKeyPair() else generateEd25519KeyPair()
+    val javaPrivateKey = getJavaKey(connectionKeys.first)
+    val sslContext = (
+        if (ch.isInitiator)
+            SslContextBuilder.forClient().keyManager(javaPrivateKey, listOf(buildCert(localKey, connectionKeys.first)))
+        else
+            SslContextBuilder.forServer(javaPrivateKey, listOf(buildCert(localKey, connectionKeys.first)))
+        )
+        .protocols(listOf("TLSv1.3"))
+        .ciphers(listOf("TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256"))
+        .clientAuth(ClientAuth.REQUIRE)
+        .trustManager(Libp2pTrustManager(expectedRemotePeer))
+        .sslContextProvider(BouncyCastleJsseProvider())
+        .applicationProtocolConfig(
+            ApplicationProtocolConfig(
+                ApplicationProtocolConfig.Protocol.ALPN,
+                ApplicationProtocolConfig.SelectorFailureBehavior.FATAL_ALERT,
+                ApplicationProtocolConfig.SelectedListenerFailureBehavior.FATAL_ALERT,
+                muxerIds.plus("libp2p") // early muxer negotiation
```

Comment:
Why do we need adding `"libp2p"` here? 
Just doing some minor refactoring and curious if it would be the same effect if we don't add this value?

Reply:
This is for compatibility with clients that do not support early muxer negotiation (they just send "libp2p"). So definitely necessary.